common/src/lib.rs

```diff
@@ -19,22 +83,23 @@ pub struct BlockHeader {
     pub timestamp: u64,
     /// The Merkle root of transactions.
     pub tx_merkle_root: Hash256,
-    /// The Merkle root of the state.
+    /// The Merkle root of the non-essential state.
     pub state_merkle_root: Hash256,
-    /// The set of validator & voteing power for the next block.
-    ///
-    /// The order here is same as the order of leaders in each round.
-    pub validator_set: Vec<(PublicKey, u64)>,
+    /// The essential state.
+    pub essential_state: EssentialState,
 }
 
 impl BlockHeader {
     pub fn hash(&self) -> Hash256 {
-        unimplemented!()
+        Hash256::hash(serde_json::to_vec(self).unwrap())
     }
 
     /// Verifies whether the given block header is a valid successor of this block.
     ///
-    /// Note that you still need to verify the block body and the finalization proof.
+    /// Note that you still need to verify
```

Comment:
no function prototypes for 1, 3 yet?

Reply:
nope. It should be in the impl. of simperby node

Reply:
ok